pkg/util/fxutil/args.go

```diff
@@ -0,0 +1,80 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package fxutil
+
+import (
+	"reflect"
+
+	"go.uber.org/fx"
+)
+
+var errorInterface = reflect.TypeOf((*error)(nil)).Elem()
+
+// delayedFxInvocation delays execution of a function, while allowing
+// Fx to provide its arguments.
+type delayedFxInvocation struct {
+	fn    interface{}
+	ftype reflect.Type
+	args  []reflect.Value
+}
+
+// newDelayedFxInvocation creates a new delayedFxInvocation wrapping the given
```

Comment:
I have tried to find an alternative implementation. It might more "fx" idiomatic but not really simpler. Feel free to skip it.

```go
var errorInterface = reflect.TypeOf((*error)(nil)).Elem()

func FnToStartHook(fn interface{}) fx.Option {
	ftype := reflect.TypeOf(fn)
	if ftype == nil || ftype.Kind() != reflect.Func {
		panic("FnToStartHook requires a function as its first argument")
	}

	in := []reflect.Type{reflect.TypeOf((*fx.Lifecycle)(nil)).Elem()}
	for i := 0; i < ftype.NumIn(); i++ {
		in = append(in, ftype.In(i))
	}

	if ftype.NumOut() > 1 || (ftype.NumOut() == 1 && !ftype.Out(0).Implements(errorInterface)) {
		panic("delayedFxInvocation function must return error or nothing")
	}

	out := make([]reflect.Type, 0)
	if ftype.NumOut() == 1 {
		out = append(out, ftype.Out(0))
	}

	newFnSig := reflect.FuncOf(in, out, false)
	wrapperFn := reflect.MakeFunc(
		newFnSig,
		func(args []reflect.Value) []reflect.Value {
			lifecycle := args[0].Interface().(fx.Lifecycle)
			lifecycle.Append(fx.Hook{
				OnStart: func(ctx context.Context) error {
					res := reflect.ValueOf(fn).Call(args[1:])

					// and return an error if the function returned any non-nil value
					if len(res) > 0 && !res[0].IsNil() {
						err := res[0].Interface().(error)
						return err
					}

					return nil
				},
			})
			return []reflect.Value{reflect.Zero(errorInterface)}
		})

	// fx.Invoke that function to capture the args at startup
	return fx.Invoke(wrapperFn.Interface())
}

```

Reply:
Nice!  This still falls to the partial-order issue I mentioned earlier: it may be that other `Start` hooks have not yet been called when this one is called, as the dependency order only produces a partial order on hooks.